aws/logs_monitoring/lambda_function.py

```diff
@@ -61,78 +72,198 @@
+        except Exception as e:
+            # most likely a network error, reset the connection
+            self._reset()
+            raise RetriableException()
+
+
+class DatadogHTTPClient(object):
+    """
+    Client that sends a batch of logs over HTTP.
+    """
+
+    _POST = "POST"
+    _HEADERS = {"Content-type": "application/json"}
+
+    def __init__(self, host, api_key, scrubber, timeout=10):
+        self._url = "https://{}/v1/input/{}".format(host, api_key)
+        self._scrubber = scrubber
+        self._timeout = timeout
+        self._session = None
+
+    def connect(self):
+        self._session = requests.Session()
+        self._session.headers.update(self._HEADERS)
+
+    def close(self):
+        self._session.close()
+
+    def send(self, logs, metadata):
+        """
+        Sends a batch of log, only retry on server and network errors.
+        """
+        try:
+            resp = self._session.post(
+                self._url,
+                data=self._scrubber.scrub(json.dumps(logs)),
+                params=metadata,
+                timeout=self._timeout,
+            )
+        except ScrubbingException as e:
+            raise Exception("could not scrub the payload")
+        except Exception as e:
+            # most likely a network error
+            raise RetriableException()
+        if resp.status_code >= 500:
+            # server error
+            raise RetriableException()
+        elif resp.status_code >= 400:
+            # client error
+            raise Exception(
+                "client error, status: {}, reason {}".format(
+                    resp.status_code, resp.reason
+                )
+            )
+        else:
+            # success
+            return
+
+
+class DatadogBatcher(object):
+    def __init__(self, max_log_size_bytes, max_size_bytes, max_size_count):
+        self._max_log_size_bytes = max_log_size_bytes
+        self._max_size_bytes = max_size_bytes
+        self._max_size_count = max_size_count
+
+    def _sizeof_bytes(self, log):
+        return len(json.dumps(log).encode("UTF-8"))
+
+    def batch(self, logs):
+        """
+        Returns an array of batches.
+        Each batch contains at most max_size_count logs and
+        is not strictly greater than max_size_bytes.
+        All logs strictly greater than max_log_size_bytes are dropped.
+        """
+        batches = []
+        batch = []
+        size_bytes = 0
+        for log in logs:
+            log_size_bytes = self._sizeof_bytes(log)
+            if (
+                len(batch) >= self._max_size_count
+                or size_bytes + log_size_bytes > self._max_size_bytes
+            ) and len(batch) > 0:
+                batches.append(batch)
+                batch = []
+                size_bytes = 0
+            if log_size_bytes <= self._max_log_size_bytes:
+                batch.append(log)
+                size_bytes += log_size_bytes
+        if len(batch) > 0:
```

Comment:
here we are missing some edge cases, like what happens when a single log is bigger than the batch size

Reply:
this is caught here https://github.com/DataDog/datadog-serverless-functions/pull/87/files#diff-618731ddf1c446ef45d09613cb4b189bR248, the message is drop, it's also explained here https://github.com/DataDog/datadog-serverless-functions/pull/87/files#diff-618731ddf1c446ef45d09613cb4b189bR234 but I can add an additional comment